### PR TITLE
Thread safety for all runtime hooks

### PR DESCRIPTION
**Pull request by gadmm**
This implements the thread-safety of hooks via:
- making global variables atomic when needed,
- documenting thread-safety assumptions.

Hooks used by users are made atomic, the other ones are documented that they do not change when several domains are running.

For thread-safety, the initialization of the `Thread` module fails if more than one domain is running.

This PR restores GC timing hooks in multicore.

This PR is best reviewed commit-per-commit.

(this could interest @kayceesrk, @jhjourdan, @Engil for reviewing)

### CODE REVIEW

**Review by xavierleroy — COMMENTED** [review]
On the one hand, the change makes sense, and I like the use of `atomic_exchange` to chain hooks.

On the other hand, I would expect hooks to be set during program initialization, before domains are spawned.  This is definitely the case for the scan_roots_hook set by the Theads library.  So, the alternative would be a stern documentation comment "don't change when multiple domains are running".

One way to decide could be to do an OPAM-wide grep for these hooks, to see how much code we're breaking in either of the two approaches above.

**Comment by gadmm** [comment]
This was actually decided according to a opam-wide grep of these hooks.

Flowtype is the only user of GC timing hooks on opam (at least according to my very old opam archive), they add and remove them on the fly, like a switch. Atomic ensures this does not become UB and I expect that change is backwards-compatible (C atomics used as r/l-values defaulting as SC read/writes).

Regarding scan_roots_hook, this is internal so the (very specialized) users can be asked to adapt their code to conform to the new way of doing. The signature has changed anyway and users might already get an error at compilation (if not then one could change the name of the hook to force an error). Our experience with ocaml-boxroot showed us that it was much simpler to not have to reason about initialization order. Currently, the library initialization is automatic and delayed until the very first allocation of a boxroot. For that we need it atomic and expect to do an exchange. In addition the previous non-atomic exchange will keep working in single-domain (modulo the change in signature).

**Comment by xavierleroy** [comment]
> Flowtype is the only user of GC timing hooks on opam (at least according to my very old opam archive), they add and remove them on the fly, like a switch. Atomic ensures this does not become UB and I expect that change is backwards-compatible (C atomics used as r/l-values defaulting as SC read/writes).

Oh, I forgot about this last bit (atomics can still be read and written without using stdatomic.h functions).  Then we're not breaking anything, and this PR is fine.

File: runtime/caml/misc.h
Comment (gasche) [inline runtime/caml/misc.h]: Is there a strong reason for making this a macro? I would just use a static inline, taking the hook address as parameter.

File: runtime/finalise.c
Comment (gasche) [inline runtime/finalise.c:148]: I think the hook should be called from within the test, not outside it.

**Review by gasche — COMMENTED** [review]
This looks good. See review comments for a minor bug in the placement of `caml_finalise_begin_hook`.

File: otherlibs/systhreads/st_stubs.c
Comment (kayceesrk) [inline otherlibs/systhreads/st_stubs.c:418]: Good catch.

File: runtime/caml/roots.h
Comment (kayceesrk) [inline runtime/caml/roots.h:26]: `camlatomic.h` does not assume that `stdatomic.h` to be present. See https://github.com/ocaml/ocaml/blob/trunk/runtime/caml/camlatomic.h. Is this breaking this assumption?

File: runtime/roots.c
Comment (kayceesrk) [inline runtime/roots.c]: `camlatomic.h` does not assume that `stdatomic.h` to be present. See https://github.com/ocaml/ocaml/blob/trunk/runtime/caml/camlatomic.h. Is this breaking this assumption?

File: otherlibs/systhreads/thread.ml
Comment (kayceesrk) [inline otherlibs/systhreads/thread.ml:89]: Is there a reason for switching the order of these functions?

File: runtime/misc.c
Comment (kayceesrk) [inline runtime/misc.c]: See earlier comments about the portability of `_Atomic`.

**Comment by abbysmal** [comment]
I think the changes make sense and I don't have more comments to make outside of what was already commented here.

One thing I noticed, is that `caml_domain_start_hook` is actually unused, I think it is a leftover of a previous implementation of a thread-safe `str`, which has been changed since then to match a better alternative from `trunk`.

It may be worth removing that hook if it is unused by the compiler distribution itself?
(unless there's a specific use that comes to mind with features not yet backported from trunk)

File: runtime/finalise.c
Comment (gadmm) [reply runtime/finalise.c:148]: Well-spotted, thanks

File: runtime/caml/misc.h
Comment (gadmm) [reply runtime/caml/misc.h]: Right, it's better to avoid macros when we can.

File: otherlibs/systhreads/thread.ml
Comment (gadmm) [reply otherlibs/systhreads/thread.ml:89]: Yes, thread_initialize can now raise an exception so we do it before any other modification to the program state. Since after the call you are sure that you are running on the only domain, this does not change behaviour.

File: runtime/caml/roots.h
Comment (gadmm) [reply runtime/caml/roots.h:26]: Indeed, something has to be done.
The main issue is gcc 4.8 support which has no equivalent to the `_Atomic` keyword, and instead has atomic operations taking arbitrary pointer types.
I see two ways to fix this. The simplest one is to introduce a `CamlAtomic` which expands to `_Atomic` or nothing, with the following two changes:
- Under gcc 4.8 define atomics as aliases for their underlying type, instead of structs. The current definition as a struct might catch more type errors, but it might be less important to support the developer experience under gcc 4.8, than supporting legacy systems running gcc 4.8 themselves.
- Don't treat C++ specially, you can include `<stdatomic.h>` from C++. Is there a reason for redefining `stdatomic.h` primitives from the C++ `atomic` header currently?

There is a second solution which is a bit more complicated so I would try to avoid it at first.

File: otherlibs/systhreads/thread.ml
Comment (kayceesrk) [reply otherlibs/systhreads/thread.ml:89]: Thanks.

File: runtime/caml/roots.h
Comment (gadmm) [reply runtime/caml/roots.h:26]: Is there a CI that currently tests gcc 4.8? The `_Atomic` keyword is used liberally in `lfskiplist.h`; according to issues on other github projects this is not supposed to work with gcc 4.8.

Comment (xavierleroy) [reply runtime/caml/roots.h:26]: I think we're moving towards not supporting GCC 4.8 and requiring proper C11 support, see #10940.  MSVC support will be another can of worms... to be opened much later.

For the time being, let's use `_Atomic` types freely, we have more urgent concerns.

Comment (xavierleroy) [reply runtime/caml/roots.h:26]: > Is there a CI that currently tests gcc 4.8? 

One of the Jenkins CI machines runs CentOS 7, which has the problematic GCC 4.8 if I remember correctly, but this part of the CI is currently turned off waiting for #10913 to be resolved.

Comment (gadmm) [reply runtime/caml/roots.h:26]: OK. Just in case, I have a fix already written.

**Comment by gadmm** [comment]
Thanks, I think it is ready.

File: otherlibs/systhreads/st_stubs.c
Comment (abbysmal) [reply otherlibs/systhreads/st_stubs.c:418]: I have been thinking about this line a bit, and it made me wonder about a probably wider area of concern: what happens in the case of another library linked in, with a top-level side effect spawning a domain?
(`caml_thread_initialized` being called from the toplevel of the thread module)

I do not think we have any way to assert that this would happen beforehand, but I may be wrong and there may be subtleties I'm missing.
I do not think that top-level domain spawning would be good practice either, however people are bound to do weird things and I'm thinking about the possible interactions and results.
Maybe it is not worth pondering upon this too much?

Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c:418]: I wondered the same; are we sure that we control initialization order enough or is there a reason one should want to initialize Threads while other domains are running? In any case, this PR goes for the simplest solution. The more general problem might require a much more complicated solution and it is best addressed separately. Should an issue be opened for this?

Comment (abbysmal) [reply otherlibs/systhreads/st_stubs.c:418]: I think there's something generally worrying if a domain were to be spawned before we get to this codepath, because the domain in question would not go through `caml_thread_initialize_domain` (since we couldn't register the callback with `Domain.at_startup`, as done in `thread.ml`.).

I don't think `systhreads` should be setting up in the presence of Multiple domains, and the current code (as of trunk) is written under the assumption that it is done before any other domains is spawned.
In that regard I think your solution fixes a nefarious scenario and we should go with it.

However we should definitely open a issue to discuss more this problem.
I see this being a problem with user-code also relying on `Domain.at_startup` and being unexpectedly hit with a top-level `Domain.spawn`. Maybe it is fine? Maybe we should add restrictions somewhere in the pipeline? (restricting top-level domain spawn?). Likely we should document and describe best-practices.

File: runtime/caml/roots.h
Comment (kayceesrk) [reply runtime/caml/roots.h:26]: I’m ok with requiring proper C11 support.

File: otherlibs/systhreads/st_stubs.c
Comment (xavierleroy) [reply otherlibs/systhreads/st_stubs.c:418]: Yes, the problematic case is a library that 1- is linked before the Threads library in linking order, and 2- whose initialization code spawns one or several domains.  1- is not very likely, as Threads is a "base" library that is used by many others, and 2- is very bad programming in my opinion.  So it's safer to have a dynamic check, knowing that it should never trigger in real use.

I'm afraid there is no way to prevent module initialization code to spawn domains, and it's a legitimate thing to do for the initialization code of the final module of a standalone application !  However, modules that we put in libraries tend to have well-behaved initialization code with no observable effects...

Comment (gasche) [reply otherlibs/systhreads/st_stubs.c:418]: I don't see anything conceptually wrong with library initialization code that would, say, decide build an immutable representation of `/usr/share/zoneinfo` at startup, traversed and processed in parallel (so: spawn several domains using a `parmap`-style helper, and then join on them again... unless the `parmap` library decides to keep them in a global thread pool for later reuse). Of course, the library authors could do this lazily instead.

On the other hand, I guess there isn't any way to prevent this, now that we don't use the `-thread` option anymore (which could ensure that threads are initialized at runtime startup) but just rely on linking order for the Threads module. (We could have a special "I'm really important, please run me first" flag, or just hope that affected users have enough control on their build system to force Threads-related module to go first in command-line order.)

Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c:418]: @Engil Another thing I saw is that loading Thread overwrites the default hooks that deal with the backup thread. How is the backup thread woken up with Threads currently? E.g. imagining the case where all threads are in a blocking section and you need a backup thread to serve a STW request.

Maybe there is there an opportunity to simplify everything and have Thread automatically initialized (before any OCaml code runs). What use-case serves _not_ having (sys)Threads?

Anyway let's split this discussion from this PR.

**Comment by gadmm** [comment]
CI is green now.

File: otherlibs/systhreads/st_stubs.c
Comment (abbysmal) [reply otherlibs/systhreads/st_stubs.c:418]: @gadmm we can discuss about this in more details in another issue (or someplace else), the gist of it though is that if you look in `st_posix.h`, you will find that the `st_masterlock_release`/`st_masterlock_acquire` cycle handles the duty of messaging or not the backup thread depending on whether we have waiters or not.
There some comments around the functions dealing with the release/acquire cycle, feel free to page me if you need more details.

Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c:418]: Alright, I missed that. Looks like what I expected, thanks.

**Comment by abbysmal** [comment]
LGTM, I think we have enough approval now to merge this?

**Comment by gasche** [comment]
Before a PR is mergd, we need at least one maintainer to "approve" it through the review interface. They are taking responsibility for the merge :-) (But sometimes discussion continue after approval, and it's good form to wait until a consensus is reached again.)  If you are thinking of merging it, you should approve it first.

Then there is the matter of the commit history: do we want to "merge" or to "squash and merge"? If the commit history is clean with separated commits, "merge" is to be preferred. If the patch remains small and meaningful as a single commit, "squash" is fine. In the present case you are in-between the two: a prefix of the history is clean (indicating the authors' intent to maybe have their work split in separate commits), and some of the review commits should probably be fixuped/squashed. In this case I think it's best to ask the author if they intend to cleanup the history, or if they are fine with a full squash. @gadmm?

**Comment by gasche** [comment]
(The "approval" part of this is documented in [CONTRIBUTING.md#Workflow](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#workflow), but I guess it could be clearer (and I'm not convinced by the phrasing of the "Allow edits from maintainers" paragraph. The "commit history" part is touched by the [Clean patch series](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#clean-patch-series) section, but that one only talks about obsessively cleaning up commit history, I guess I wrote it, and it doesn't mention the easy way out. Probably for the best.)

**Comment by gadmm** [comment]
I usually clean-up my histories before asking for a merge and expect no squash. But given your question we seem to disagree that it is clean so I am going to squash a few more commits now.

**Comment by gasche** [comment]
I think that "fix clang build" and "gasche's review comments" are not desirable in the history. Sometimes it's okay to have a few undesirable commits, but here the intermediate state is broken for some compilers or (very slightly) buggy, it's not an ideal place to end up during a bisection for example. (In practice bisections skip over merge commits completely I think, so thankfully bisecting in the middle of a PR is not so common, but still.)

File: otherlibs/systhreads/st_stubs.c
Comment (xavierleroy) [reply otherlibs/systhreads/st_stubs.c:418]: > I don't see anything conceptually wrong with library initialization code that would, say, decide build an immutable representation of /usr/share/zoneinfo at startup, traversed and processed in parallel (so: spawn several domains using a parmap-style helper

One thing that is wrong here is the inability to handle errors: if something goes wrong while reading and preprocessing /usr//share/zoneinfo, an exception will be raised that no user code can handle.  That's one of the many reasons why library initialization code should perform very little computation, leaving heavy computation (e.g. that can fail or that can take significant time) to explicit `init` functions.

**Review by xavierleroy — APPROVED** [review]
This PR was reviewed extensively, and the commit history is now clean as a whistle, so let me merge ASAP.

**Pull request merged**
